### PR TITLE
refactor: Run doctor command

### LINKED ISSUE #1868: Structural output - RunDoctorAndroid
# Plan

According to the specification described in epic #1728, implement the structural output for the command and move the console logging to layer CLI.

# Command

`RunDoctorAndroid`

### PR DESCRIPTION
**Pull request by pawelpasterz**
Fixes #1868 #1869

There is no better way to test it than breaking the config yml. Then run:
* `flank firebase test android doctor -c [path to your yml]`
* `flank firebase test ios doctor -c [path to your yml]`
* add `--fix`
* 🎉

### CODE REVIEW

File: test_runner/src/main/kotlin/ftl/doctor/Doctor.kt [hunk 1]
```diff
@@ -5,43 +5,49 @@ import com.google.common.annotations.VisibleForTesting
...
+    }.filter { it.second.isNotEmpty() }.toMap(),
+    invalidDevices = validateDevices().orEmpty()
+)
 
-private fun JsonNode.validateTopLevelKeys(args: IArgs.ICompanion) =
+private fun JsonNode.validateTopLevelKeys(args: IArgs.ICompanion): List<String> =
     (parseArgs().keys - args.validArgs.keys)
         .takeIf { it.isNotEmpty() }
-        ?.let { unknownKeys -> "Unknown top level keys: $unknownKeys\n" }
+        ?.map { it }
```
Comment (piotradamczyk5) [inline test_runner/src/main/kotlin/ftl/doctor/Doctor.kt:50]: what is the purpose of this `map`?

File: test_runner/src/main/kotlin/ftl/domain/DoctorErrors.kt [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+        else buildString {
+            parsingErrors.forEach { appendLine(it) }
+            invalidDevices
+                .map { "Warning: Version should be string $GCLOUD_NODE -> $DEVICES_NODE[${it.model}] -> $VERSION_NODE[${it.version}]" }
+                .forEach { appendLine(it) }
+            if (topLevelUnknownKeys.isNotEmpty()) appendLine("Unknown top level keys: $topLevelUnknownKeys")
+            nestedUnknownKeys.forEach { (topLevelKey, keys) ->
+                appendLine("Unknown keys in $topLevelKey -> $keys")
+            }
+        }.trim()
```
Comment (piotradamczyk5) [inline test_runner/src/main/kotlin/ftl/domain/DoctorErrors.kt:33]: it should be an extension, `data class` should not have functions (linter should complain also)

Comment (pawelpasterz) [reply test_runner/src/main/kotlin/ftl/domain/DoctorErrors.kt:33]: Good point 👍🏻 Thanks

File: test_runner/src/main/kotlin/ftl/doctor/Doctor.kt (hunk 1 above)
Comment (pawelpasterz) [reply test_runner/src/main/kotlin/ftl/doctor/Doctor.kt:50]: That is an excellent question...and I don't know 😅 will be removed, Thanks!

**Review by piotradamczyk5 — CHANGES_REQUESTED** [review]
some problem with double printing under investigation

File: test_runner/src/main/kotlin/ftl/doctor/Doctor.kt (hunk 1 above)
Comment (adamfilipow92) [reply test_runner/src/main/kotlin/ftl/doctor/Doctor.kt:50]: We expecting ```List``` as result of function but ```(parseArgs().keys - args.validArgs.keys)``` returns ```Set```. I will change this map to ```.toList()```

**Review by piotradamczyk5 — APPROVED** [review]

**Review by piotradamczyk5 — APPROVED** [review]

**Review by jan-gogo — CHANGES_REQUESTED** [review]
`DoctorErrors.kt` cannot be placed directly inside `domain` package because it's not a top-level function or scope. But as I see `RunDoctorAndroid` and `RunDoctorIos` looks similar, the only difference is the first argument for `validateYaml`. So my proposition is to merge `RunDoctorIos` and `RunDoctorAndroid` into one `RunDoctor` create parameterized `RunDoctor.invoke(args: IArgs.ICompanion)`, and put `DoctorErrors` inside `RunDoctor` under `Errors` name, so in result there will be a `RunDoctor.Errors` structure. Be aware that `RunDoctor.Errors.summary()` mapper is a part of `presentation.cli`, not a `domain`.

**Review by Sloox — APPROVED** [review]

File: test_runner/src/main/kotlin/ftl/domain/RunDoctor.kt [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+    if (fix) processValidation(Paths.get(configPath))
+    validationResult.out()
+    if (!validationResult.isEmpty()) throw YmlValidationError()
+}
+
+fun RunDoctor.Error.isEmpty() =
+    parsingErrors.isEmpty() && topLevelUnknownKeys.isEmpty() &&
+        nestedUnknownKeys.isEmpty() && invalidDevices.isEmpty()
+
+operator fun RunDoctor.Error.plus(right: RunDoctor.Error) = RunDoctor.Error(
```
Comment (jan-gogo) [inline test_runner/src/main/kotlin/ftl/domain/RunDoctor.kt:43]: This is not a top-level function, it shouldn't be here.

File: test_runner/src/main/kotlin/ftl/domain/RunDoctor.kt [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+    val validationResult = validateYaml(args, path = Paths.get(configPath))
+    if (fix) processValidation(Paths.get(configPath))
+    validationResult.out()
+    if (!validationResult.isEmpty()) throw YmlValidationError()
+}
+
+fun RunDoctor.Error.isEmpty() =
+    parsingErrors.isEmpty() && topLevelUnknownKeys.isEmpty() &&
+        nestedUnknownKeys.isEmpty() && invalidDevices.isEmpty()
+
```
Comment (jan-gogo) [inline test_runner/src/main/kotlin/ftl/domain/RunDoctor.kt:42]:
```suggestion
```
This function is redundant due to `error == RunDoctor.Error.EMPTY` operation.

**Review by jan-gogo — CHANGES_REQUESTED** [review]

File: test_runner/src/main/kotlin/ftl/domain/RunDoctor.kt (hunk 3 above)
Comment (adamfilipow92) [reply test_runner/src/main/kotlin/ftl/domain/RunDoctor.kt]: Thanks! Fixed! 👍

**Pull request merged**